passage/src/main/java/id/passage/android/PassageHosted.kt

```diff
@@ -0,0 +1,177 @@
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.withContext
+import okhttp3.MediaType.Companion.toMediaType
+import okhttp3.OkHttpClient
+import okhttp3.Request
+import okhttp3.RequestBody.Companion.toRequestBody
+import java.net.URLEncoder
+import java.security.MessageDigest
+import java.security.SecureRandom
+import java.util.Base64
+
+internal class PassageHosted {
+    internal companion object {
+        private var verifier = ""
+        private var state = ""
+        private const val CODE_CHALLENGE_METHOD = "S256"
+        private const val SECRET_STRING_LENGTH = 32
+        private val basePathOIDC = "https://${Passage.authOrigin}"
+        private val appId = Passage.appId
+        private val packageName = Passage.packageName
+
+        internal fun openChromeTab(activity: Activity) {
+            val redirectUri = "$basePathOIDC/android/$packageName/callback"
+            state = getRandomString()
+            val randomString = getRandomString()
+            verifier = getRandomString()
+            val codeChallenge = sha256Hash(randomString)
+            val newParams =
+                listOf(
+                    "client_id" to appId,
+                    "redirect_uri" to redirectUri,
+                    "state" to state,
+                    "code_challenge" to codeChallenge,
+                    "code_challenge_method" to CODE_CHALLENGE_METHOD,
+                    "scope" to "openid",
+                    "response_type" to "code",
+                ).joinToString("&") { (key, value) ->
+                    "$key=${URLEncoder.encode(value, "UTF-8")}"
+                }
+            val url = "$basePathOIDC/authorize?$newParams"
+            val intent = CustomTabsIntent.Builder().build()
+            intent.launchUrl(activity, Uri.parse(url))
+        }
+
+        private fun getRandomString(): String {
+            val digits = '0'..'9'
+            val upperCaseLetters = 'A'..'Z'
+            val lowerCaseLetters = 'a'..'z'
+            val characters = (digits + upperCaseLetters + lowerCaseLetters)
+                .joinToString("")
+            val random = SecureRandom()
+            val stringBuilder = StringBuilder(SECRET_STRING_LENGTH)
+            for (i in 0 until SECRET_STRING_LENGTH) {
+                val randomIndex = random.nextInt(characters.length)
+                stringBuilder.append(characters[randomIndex])
+            }
+            return stringBuilder.toString()
+        }
+
+        private fun sha256Hash(randomString: String): String {
+            val bytes = randomString.toByteArray()
+            val md = MessageDigest.getInstance("SHA-256")
+            val digest = md.digest(bytes)
+            return Base64.getUrlEncoder().withoutPadding().encodeToString(digest)
+        }
+
+        internal suspend fun finishOIDC(
+            activity: Activity,
+            code: String,
+            clientSecret: String,
+            state: String,
+        ): AuthResult? {
```

Comment:
Can we make this expected return type non-optional? If this method does NOT return an auth result, I would expect it to throw an exception?
```suggestion
        ): AuthResult {
```